File: leetcode/src/longest-substring-without-repeating-characters.cpp
#include<unordered_map>
#include <algorithm>
using namespace std;
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        auto p1 = 0;
        auto p2 = 0;
        int max_len = 0;
        if (s.size() <= 1) {
            return s.size();
        }
        unordered_map<char, int> chars;
        chars[s[0]] = 0;
        while (p2+1 != s.size()) {
            char nextchar = s[p2+1];
            if (chars.find(nextchar) != chars.end() && chars[nextchar] >= p1) {
                p1 = chars[nextchar]+1;
            }
            chars[nextchar] = p2+1;
            ++p2;
            max_len = max(max_len, p2-p1+1);
        }
        return max_len;
    }
};

File: leetcode/src/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxLength = 0
        if (len(s) <= 1):
            return len(s)
        p1 = 0
        p2 = 0
        chars = {s[0]: 0}
        while (p2 != len(s)-1):
            nextChar = s[p2+1]
            if (nextChar in chars and chars[nextChar] >= p1 and chars[nextChar] <= p2):
                p1 = chars[nextChar]+1
            chars[nextChar] = p2+1
            p2 += 1
            maxLength = max(maxLength, p2-p1+1)
        
        return maxLength